components/desktop/freerdp/Makefile

```diff
@@ -0,0 +1,118 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"). You may
+# only use this file in accordance with the terms of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source. A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Michal Nowak
+#
+
+PREFERRED_BITS=		64
+
+include ../../../make-rules/shared-macros.mk
+
+COMPONENT_NAME=		FreeRDP
+COMPONENT_VERSION=	2.0.0
+COMPONENT_COMMIT=	e873e4f0042dd157f68ae99adb4dc9dcd167ca29
+COMPONENT_SUMMARY=	Remote Desktop Viewer Client
+COMPONENT_PROJECT_URL=	http://www.freerdp.com
+COMPONENT_FMRI=		desktop/remote-desktop/freerdp
+COMPONENT_CLASSIFICATION=	Applications/Internet
+COMPONENT_SRC=		$(COMPONENT_NAME)-$(COMPONENT_COMMIT)
+COMPONENT_ARCHIVE=	$(COMPONENT_SRC).zip
+COMPONENT_ARCHIVE_URL=	https://github.com/FreeRDP/FreeRDP/archive/$(COMPONENT_COMMIT).zip
+COMPONENT_ARCHIVE_HASH=	\
+	sha256:12374e828745e13f707844c4fc4607094f16513b8b94e6728883690fdb9cdc8e
+COMPONENT_LICENSE=	Apache v2.0
+COMPONENT_LICENSE_FILE=	LICENSE
+
+include $(WS_MAKE_RULES)/prep.mk
+include $(WS_MAKE_RULES)/cmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+# https://github.com/FreeRDP/FreeRDP/issues/5504
+CFLAGS += -pthread
```

Comment:
hm... Can you rise a question in \#illumos, I do not understand this. libpthread is just libc filter...

Reply:
Perhaps, gcc adds some defines when -pthread is used...

Reply:
I think for us it essentially boils down to:
```
CFLAGS += -D_REENTRANT -D_PTHREADS
LDFLAGS += -lpthread -ldl -lc
```

Per https://stackoverflow.com/a/23251828:
```patch
$ gdiff -pruN1 dm.nopthread.txt dm.pthread.txt 
--- dm.nopthread.txt	2019-07-30 10:20:41.275025417 +0000
+++ dm.pthread.txt	2019-07-30 10:20:33.776335769 +0000
@@ -66,2 +66,3 @@
 #define __LDBL_HAS_QUIET_NAN__ 1
+#define _REENTRANT 1
 #define __GNUC__ 6
@@ -144,2 +145,3 @@
 #define __INT_LEAST16_MAX__ 0x7fff
+#define _PTHREADS 1
 #define __DEC64_MANT_DIG__ 16
```

```
$ gcc -dumpspecs | grep '%{pthread'
%{!symbolic:     %{pthreads|pthread:-lpthread}      %{p|pg:-ldl} -lc}
%{pthreads|pthread:-D_REENTRANT -D_PTHREADS}
```

Reply:
https://github.com/FreeRDP/FreeRDP/issues/5504#issuecomment-516314097

Can it be (and should it be?) that our OpenSSL is without `-pthreads`?

Reply:
We have `CONFIGURE_OPTIONS += threads` for openssl... 
Interesting, if it's question of defines vs question of libs. If you compile just with -D_REENTRANT, does it behave sane?

Reply:
That alone works fine.

Reply:
Can you rise this question at illumos-dev what heritage is there? I mean, why it's needed at all...

Reply:
And, more funny, defining _REENTRANT, enables posix variants of some functions (like readdir_r()... Does it just have an effect of XOPEN_SOURCE=600 ?

Reply:
That fails early with:
```
FreeRDP-e873e4f0042dd157f68ae99adb4dc9dcd167ca29/winpr/include/winpr/winsock.h:318:48: error: unknown type name 'u_long'
 WINPR_API int _ioctlsocket(SOCKET s, long cmd, u_long* argp);
                                                ^~~~~~
```

Reply:
> Can you rise this question at illumos-dev what heritage is there? I mean, why it's needed at all...

I will do later.

Reply:
Try CPPFLAGS+=$(CPP_XPG6MODE)

Reply:
It would be good to make it default (at least for oi-userland) one day :)

Reply:
`CFLAGS += $(XPG6MODE)` does work.

Reply:
> `CFLAGS += $(XPG6MODE)` does work.

Should I keep it or is any other variant preferred?

Reply:
CFLAGS += $(XPG6MODE) is fine.